### ISSUE TITLE
What it tells, when SSD Loss is exploding, while val_loss still decreases

### ISSUE THREAD

**Issue by pinkerltm** [pierluigiferrari/ssd_keras#118]
During Training, after epoch ~30 my training loss is exploding. Any idea what might cause this? 

![image](https://user-images.githubusercontent.com/6107708/39762872-0a357fc8-52dc-11e8-8c2d-16e0c7a5894f.png)

**Comment by pierluigiferrari** [comment]
1. Does the loss before the highlighted part behave normally? It's not clear on the huge scale of the loss plot.
2. Did this happen only once or did it happen repeatedly?
3. The accuracy graphs don't look right, either. Both the training and validation accuracies first increase and then keep decreasing? Something can't be right there.

I don't have an explanation for those loss spikes, but it would be interesting to see exactly what mini batch is the first to cause a drastic loss increase. The easiest case would be if something were wrong with the data itself (even though it's unclear why this would only show after 38 epochs, but then again I also don't know what one epoch is in your training). The less easy case would be if there was nothing obviously wrong with the data in the mini batch, because then the next steps would be to investigate what's going on inside the loss function for that particular mini batch.

**Comment by pinkerltm** [comment]
ad 1. Yes, it is following a hyperbolae decrease from up over 100 down to around 14...it seemed to do very well until then

ad 2. It happens the first time and I thought that I have now a hyperparameter setting which should at least g
converge to a robust but not ideal detector

ad 3. There are 80 classes (just like COCO) so accuracy should at least be significantly higher than 1/80 shouldn't it? otherwise its just a guess. 

My Batch size is 16, one Epoch is 1000 steps, so it should be 16000 samples out of 5000 training images if i got it right.

Since this is my very first broader Experiment and I am still not understanding a lot of things, its possible that I am doing a lot of things wrong, but I already trained a detector on this dataset with vanilla ssd_keras code and settings which converged and had about 70% accuracy. 

I changed only a mapping from the datasets 250 different labels to only 80 which are more relevant to my use case, so that I can match belgian traffic signs to traffic signs of other countries.

Today training should finish, I will test if anything usable came out, tweak it a bit and report here further if you are interested. 

Do you have any resources on how to debug the loss function during training which I can dive into?

**Comment by pierluigiferrari** [comment]
> ad 3. There are 80 classes (just like COCO) so accuracy should at least be significantly higher than 1/80 shouldn't it? otherwise its just a guess.

Yes, but the absolute value of the accuracies is not the point here (they are both above 30 percent, so the model seems to be learning something). The point is that the development of the accuracy is weird. First it goes up very fast, then it decreases steadily. As I said, something is wrong here.

> vanilla ssd_keras code

Meaning this SSD implementation?

> I changed only a mapping from the datasets 250 different labels to only 80 which are more relevant to my use case

Meaning the model that converged at 70% accuracy was trained on the full 250 classes?

**Comment by pinkerltm** [comment]
> Meaning this SSD implementation?

Yes, just cloned your repo and trained on my data

> Meaning the model that converged at 70% accuracy was trained on the full 250 classes?

Yes, which resulted in about 55 Mio. trainable Parameters which took quite Long

ad Loss: mhm, ok... I think it will be already helpful to see the Training Batch which causes the spikes to get some ideas. I'll give it a try. Thanks.

**Comment by pierluigiferrari** [comment]
That's good news. If the training worked well with the full 250 classes then you have the first part of the solution already: There is almost certainly some problem with the reduced 80-class dataset you created. Insofar I'd also assume that it could be quite insightful already to take a look at the batch that causes the first spike.
